pkg/agent/openflow/pipeline.go

```diff
@@ -825,32 +960,185 @@ func (c *client) l3ToExternalFlows(nodeIP net.IP, localSubnet net.IPNet, outputP
+	} else if protocol == binding.ProtocolSCTP {
+		lbFlowBuilder = lbFlowBuilder.MatchSCTPDstPort(svcPort)
+	}
+	lbFlow := lbFlowBuilder.
+		MatchDstIP(svcIP).
+		MatchRegRange(int(serviceLearnReg), marksRegServiceNeedLB, serviceLearnRegRange).
+		Action().Group(groupID).
+		Cookie(c.cookieAllocator.Request(cookie.Service).Raw()).
+		Done()
+	return lbFlow
+}
+
+// endpointDNATFlow generates the flow which transforms the Service Cluster IP
+// to the Endpoint IP according to the Endpoint selection decision which is stored
+// in regs.
+func (c *client) endpointDNATFlow(endpointIP net.IP, endpointPort uint16, protocol binding.Protocol) binding.Flow {
+	ipVal := binary.BigEndian.Uint32(endpointIP)
+	unionVal := (marksRegServiceHitted << endpointPortRegRange.Length()) + uint32(endpointPort)
+	return c.pipeline[endpointDNATTable].BuildFlow(priorityNormal).
+		Cookie(c.cookieAllocator.Request(cookie.Service).Raw()).
+		MatchProtocol(protocol).
+		MatchReg(int(endpointIPReg), ipVal).
+		MatchRegRange(int(endpointPortReg), unionVal, binding.Range{0, 18}).
+		Action().CT(true, egressRuleTable, ctZone).
+		DNAT(
+			&binding.IPRange{StartIP: endpointIP, EndIP: endpointIP},
+			&binding.PortRange{StartPort: endpointPort, EndPort: endpointPort},
+		).
+		LoadToMark(serviceCTMark).
+		CTDone().
+		Done()
+}
+
+// hairpinSNATFlow generates the flow which does SNAT for Service
+// hairpin packets and loads the hairpin mark to markReg.
+func (c *client) hairpinSNATFlow(endpointIP net.IP) binding.Flow {
+	return c.pipeline[hairpinSNATTable].BuildFlow(priorityNormal).
+		Cookie(c.cookieAllocator.Request(cookie.Service).Raw()).
+		MatchProtocol(binding.ProtocolIP).
+		MatchDstIP(endpointIP).
+		MatchSrcIP(endpointIP).
+		Action().SetSrcIP(hairpinIP).
+		Action().LoadRegRange(int(marksReg), hairpinMark, hairpinMarkRange).
+		Action().GotoTable(l2ForwardingOutTable).
+		Done()
+}
+
+// serviceEndpointGroup creates/modifies the group/buckets of Endpoints. If the
+// withSessionAffinity is true, then buckets will resubmit packets back to
+// serviceLBTable to trigger the learn flow, the learn flow will then send packets
+// to endpointDNATTable. Otherwise, buckets will resubmit packets to
+// endpointDNATTable directly.
+func (c *client) serviceEndpointGroup(groupID binding.GroupIDType, withSessionAffinity bool, endpoints ...proxy.Endpoint) binding.Group {
+	group := c.bridge.CreateGroup(groupID).ResetBuckets()
+	var resubmitTableID binding.TableIDType
+	var lbResultMark uint32
+	if withSessionAffinity {
+		resubmitTableID = serviceLBTable
+		lbResultMark = marksRegServiceNeedLearn
+	} else {
+		resubmitTableID = endpointDNATTable
+		lbResultMark = marksRegServiceHitted
+	}
+
+	for _, endpoint := range endpoints {
+		endpointPort, _ := endpoint.Port()
+		endpointIP := net.ParseIP(endpoint.IP()).To4()
+		ipVal := binary.BigEndian.Uint32(endpointIP)
+		portVal := uint16(endpointPort)
+		// TODO: use goto_table instead.
```

Comment:
anything preventing us from using goto_table right away?

Reply:
I think the upstream ofnet does not support this in bucket.

Reply:
I have just tried to add a group with goto_table by using ovs-ofctl and failed. It looks like the goto_table is not allowed in a bucket's actions.
> ovs_version: "2.13.0"
> ovs-ofctl: goto_table instruction not allowed here

Removed this line.